providers/os/connection/snapshot/blockdevices.go

```diff
@@ -125,29 +182,70 @@ func (blockEntries BlockDevices) GetMountablePartitionByDevice(device string) (*
 	return &PartitionInfo{Name: devFsName, FsType: partitions[0].FsType}, nil
 }
 
+// LongestMatchingSuffix returns the length of the longest common suffix of two strings
+// and caches the result (lengths of the matching suffix) for future calls with the same string
+func LongestMatchingSuffix(s1, s2 string) int {
+	n1 := len(s1)
+	n2 := len(s2)
+
+	// Start from the end of both strings
+	i := 0
+	for i < int(math.Min(float64(n1), float64(n2))) && s1[n1-i-1] == s2[n2-i-1] {
+		i++
+	}
+
+	return i
+}
+
 // Searches for a device by name
-func (blockEntries BlockDevices) FindDevice(name string) (BlockDevice, error) {
-	log.Debug().Str("device", name).Msg("searching for device")
-	var secondName string
-	if strings.HasPrefix(name, "/dev/sd") {
-		// sdh and xvdh are interchangeable
-		end := strings.TrimPrefix(name, "/dev/sd")
-		secondName = "/dev/xvd" + end
+func (blockEntries BlockDevices) FindDevice(requested string) (BlockDevice, error) {
+	log.Debug().Str("device", requested).Msg("searching for device")
+
+	devices := blockEntries.BlockDevices
+	if len(devices) == 0 {
+		return BlockDevice{}, fmt.Errorf("no block devices found")
 	}
-	for i := range blockEntries.BlockDevices {
-		d := blockEntries.BlockDevices[i]
-		log.Debug().Str("name", d.Name).Interface("children", d.Children).Interface("mountpoint", d.MountPoint).Msg("found block device")
-		fullDeviceName := "/dev/" + d.Name
-		if name != fullDeviceName { // check if the device name matches
-			if secondName == "" || secondName != fullDeviceName {
-				continue
+
+	requestedName := strings.TrimPrefix(requested, "/dev/")
+	lmsCache := map[string]int{}
+	bestMatch := struct {
```

Comment:
can we simplify this a bit? i feel like there's a way to make this a bit more readable by iterating over all devices and storing their lms and then taking the one with the longest lms at the end, something like:
```
lmsMap := map[string]int{}
	devicesMap := map[string]BlockDevice{}
	for _, d := range devices {
		devicesMap[d.Name] = d
		if d.Name == requestedName {
			return d, nil
		}

		lms := LongestMatchingSuffix(requested, d.Name)
		for _, alias := range d.Aliases {
			// we also try all the aliases and see if those are a better match
			aliasLms := LongestMatchingSuffix(requested, alias)
			if aliasLms > lms {
				lms = aliasLms
			}
		}
		lmsMap[d.Name] = lms
	}

	device := BlockDevice{}
	for name, lms := range lmsMap {
		if lms > 0 && lms > lmsMap[device.Name] {
			device = devicesMap[name]
		}
	}

	// if we have a name, we've found a device
	if device.Name != "" {
		return devicesMap[device.Name], nil
	}
```

that avoids the need for a private struct and the iteration, starting from the 1st index

i have not actually tested this code so it could be wrong